stan/math/prim/mat/fun/value_of_rec.hpp

```diff
@@ -22,12 +22,33 @@ namespace stan {
     template <typename T, int R, int C>
     inline Eigen::Matrix<double, R, C>
     value_of_rec(const Eigen::Matrix<T, R, C>& M) {
-      using stan::math::value_of_rec;
-      Eigen::Matrix<double, R, C> Md(M.rows(), M.cols());
-      for (int j = 0; j < M.cols(); j++)
-        for (int i = 0; i < M.rows(); i++)
-          Md(i, j) = value_of_rec(M(i, j));
-      return Md;
+      Eigen::Matrix<double, R, C>
+        result(M.rows(), M.cols());
+
+      double * data_r = result.data();
+      const T * data_m = M.data();
+
+      int S = M.size();
+      for (int i=0; i < S; i++)
+        data_r[i] = value_of_rec(data_m[i]);
+      return result;
+    }
```

Comment:
Ah, I see --- the recursion isn't through container structures
(just std::vector), but also down through mixed types.  Thanks
for pointing this out.

> On Mar 3, 2016, at 3:43 PM, Rob Trangucci <notifications@github.com> wrote:
> 
> In stan/math/prim/mat/fun/value_of_rec.hpp:
> 
> > -      Eigen::Matrix<double, R, C> Md(M.rows(), M.cols());
> > -      for (int j = 0; j < M.cols(); j++)
> > -        for (int i = 0; i < M.rows(); i++)
> > -          Md(i, j) = value_of_rec(M(i, j));
> > -      return Md;
> > +      Eigen::Matrix<double, R, C>
> > +        result(M.rows(), M.cols());
> > +
> > +      double * data_r = result.data();
> > +      const T * data_m = M.data();
> > +
> > +      int S = M.size();
> > +      for (int i=0; i < S; i++)
> > +        data_r[i] = value_of_rec(data_m[i]);
> > +      return result;
> > +    }
> 
> @bob-carpenter I'm confused about the comment above about value_of_rec.
> 
> I can think of a situation where we would have:
> 
> Eigen::Matrix<fvar<var>, R, C> x
> 
> and want y where y(i, j) = x(i, j).val_.val()
> 
> —
> Reply to this email directly or view it on GitHub.
>